bots/psuteam7bot/movement.js

```diff
@@ -464,5 +378,386 @@ movement.findAdjacentBase = (self) => {
+*Input: fullMap     -  the full map, should be self.map or self.getPassableMap()
+*Output:    retVal  -   false if the map is a horizontal reflection
+*                   -   OR true if it is a vertical reflection 
+* Not accounted for: Case for when the map is both horizontally and vertically reflected
+*/
+movement.isHorizontalReflection = (fullMap) => {
+    let y = 0;
+    let x = 0;
+    const length = fullMap.length;
+    let mirror = length-1;
+
+    while(y < length)
+    {
+        while(x < mirror)
+        {
+            if(fullMap[y][x] !== fullMap[y][mirror])
+            {
+                return true
+            }
+            ++x;
+            --mirror;
+        }
+        ++y;
+        mirror = length - 1;
+        x = 0;
+    }
+    return false;
+}
+
+/**
+*Calculate and return enemy castle's potential starting location
+*Input:     myCastleLocation    -   the Castle's 'position/ location' object, should be self.me/ location of unit's base
+*           fullMap             -   the full map, Should be self.map or or self.getPassableMap()
+*Output:    RetVal  -   An {x, y} object, which should be the position of the mirror enemy castle to myCastleLocation
+*/
+movement.getMirrorCastle = (myCastleLocation, fullMap) => {
+    const {x, y} = myCastleLocation;
+    const Ax = fullMap.length - x - 1;
+    const Ay = fullMap.length - y - 1;
+    const isHorizontal = movement.isHorizontalReflection(fullMap);
+
+    if(isHorizontal)
+    {
+        return {x: x, y: Ay}
+    }
+    else
+    {
+        return {x: Ax, y: y};
+    }
+}
+
+/**
+* Calculate and return enemy castle's potential starting location
+*Input:     alliedCastleLocations       -   An array of 'position/ location' objects for the allied castles, should start with the calling castle's location followed by any allied castles
+*           fullMap                     -   the full map, Should be self.map or or self.getPassableMap()
+*Output:    RetVal  -   An array of {x, y} objects, which should be positions of enemy castles
+*/
+movement.getEnemyCastleLocations = (alliedCastleLocations, fullMap) => {
+    let enemyCastleLocations = [];
+    for(let i = 0; i < alliedCastleLocations.length; ++i)
+    {
+        enemyCastleLocations.push(movement.getMirrorCastle(alliedCastleLocations[i], fullMap));
+    }
+    return enemyCastleLocations;
+}
+
+/**
+*Calculate and return enemy castle's potential starting location
+*Input:     myCastleLocation    -   the Castle's 'position/ location' object, should be self.me/ self.base
+*           fullMap             -   the full map, Should be self.map or or self.getPassableMap()
+*Output:    RetVal  -   The diagonal mirror of myCastleLocation, should be in an enemy quadrant
+*Standalone helper function, to obtain the diagonal mirror patrol position
+*/
+movement.getDiagonalPatrolPosition = (myCastleLocation, fullMap) => {
+    const {x, y} = myCastleLocation;
```

Comment:
When do we use this? I don't see it invoked in `getAttackerPatrolRoute` which is where I assume is the most natural place to use this

Reply:
It is used when the Attacker receive a signal order from base castle for their target and long range communication is not implemented yet.